accepted/install-locations.md

```diff
@@ -0,0 +1,227 @@
+
+Unfortunately some of these components are hard to service:
+* The muxer relies on the caller to locate it, then it's shared by all versions installed in any given location. Updating it is troublesome as it typically require a system reboot (may be in use by running applications).
+* The `hostfxr` is relatively easy to service since the installer places a newer version side-by-side and relies on the host to pick the latest available version. The complexity comes from the fact that it has to work with all versions of the runtime/sdk.
+* App-contained hosts (`apphost`, `comhost`, `ijwhost`, `nethost`) these are libraries which are shipped with the apps and thus fully owned/serviced by the app. They need to know how to find install locations to locate shared frameworks.
+
+The biggest problem is with the app-contained hosts which we can't service in any way. So introducing new install types or changing the default location can be breaking for apps using old hosts.
+
+The problem already exists since the `apphost` has been shipped in .NET Core 2.1 with support for framework dependent apps. Fortunately this feature is not widely used yet.
+
+In .NET Core 3.0 the situation will be very different. Framework dependent `apphost` is the default for apps. Scenarios requiring the other hosts are also likely to be relatively common (COM, IJW, native hosting). So the hosts which we ship with .NET Core 3.0 will be spread across all machines running .NET Core (not just development but production as well) without any ability to service them directly.
+
+This means that .NET Core 3.0 is the last chance to get the algorithms for install locations modified in "breaking ways". Future release will very likely have to be 100% backward compatible in this regard.
+
+
+## Proposal for 3.0
+
+### Globally registered install location :new:
+This is by far the most common case and in 3.0 we want to provide a way to globally install to a custom location. The typical case will still be to install into the default location, but custom location should be possible. Either location should behave the same.
+
+To achieve this the install location must be recorded in some system-wide configuration in a fixed location which is well-known.
+
+* Windows - the installer will store the install location in 32-bit registry (for both 32-bit and 64-bit installations):
+  * `HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\<arch>\InstallLocation` - where `<arch>` is one of the `x86`, `x64`, `arm` or `arm64`.
+  Note that this registry key is "redirected" that means that 32-bit processes see different copy of the key then 64bit processes. So it's important that both installers and the host access only the 32-bit view of the registry.
+* Linux and macOS - the installer will store the install location in a text file
+  * `/etc/dotnet/install_location` - the file will contain a single line which is the path.
+
+The muxer (`dotnet.exe`) is registered on the system-wide `PATH` environment variable.
```

Comment:
This is not true today on linux installations. Are we going to change from using symlinks? Not opposed to the idea but would be interested to understand why we went with symlinks to begin with (decision made before I jumped into the pool) and if there are any ramifications to changing the practice.

Reply:
I think this should be clarified. It's unclear to me what "registered" means. You could argue that adding a symlink to `/usr/bin` is just the way to "register" `dotnet` on `PATH` for our Linux packages.

It might just be an implementation detail out of scope for this doc. (E.g. I believe snaps have a different mechanism to be available through `PATH`.)

Reply:
Thanks, I made this a bit more vague... and I'll leave out the specific implementation. There are probably too many variants and I have no idea how to find them all anyway.